src/test/java/bio/overture/ego/controller/PolicyControllerTest.java

```diff
@@ -234,4 +234,64 @@ public void disassociatePermissionsFromUser_ExistingEntitiesAndRelationships_Suc
     assertEquals(getResponseStatus, OK);
     assertEquals(0, getResponseJson.get("count").asInt());
   }
+
+  @Test
+  @SneakyThrows
+  public void associatePermissionsWithApplication_ExistingEntitiesButNoRelationship_Success() {
+    val policyId = entityGenerator.setupSinglePolicy("AddAppPermission").getId().toString();
+    val appId = entityGenerator.setupApplication("AppPolicy Add").getId().toString();
+
+    val response =
+        initStringRequest()
+            .endpoint("/policies/%s/permission/application/%s", policyId, appId)
+            .body(createMaskJson(READ.toString()))
+            .post();
+
+    val responseStatus = response.getStatusCode();
+    assertEquals(responseStatus, OK);
+    // TODO: Fix it so that POST returns JSON, not just random string message
+
+    val getResponse = initStringRequest().endpoint("/policies/%s/applications", policyId).get();
+
+    val getResponseStatus = getResponse.getStatusCode();
+    val getResponseJson = MAPPER.readTree(getResponse.getBody());
+    val permissionJson = getResponseJson.get("resultSet").get(0);
+
+    assertEquals(getResponseStatus, OK);
+    assertEquals(permissionJson.get("id").asText(), appId);
+    assertEquals(permissionJson.get("mask").asText(), "READ");
+  }
+
+  @Test
+  @SneakyThrows
+  public void disassociatePermissionsFromApplication_ExistingEntitiesAndRelationships_Success() {
+    val policyId = entityGenerator.setupSinglePolicy("DeleteAppPermission").getId().toString();
+    val appId = entityGenerator.setupApplication("AppPolicy Delete").getId().toString();
+
+    val response =
+        initStringRequest()
+            .endpoint("/policies/%s/permission/application/%s", policyId, appId)
+            .body(createMaskJson(WRITE.toString()))
+            .post();
+
+    val responseStatus = response.getStatusCode();
+    assertEquals(responseStatus, OK);
+    // TODO: Fix it so that POST returns JSON, not just random string message
+
+    val deleteResponse =
+        initStringRequest()
+            .endpoint("/policies/%s/permission/application/%s", policyId, appId)
+            .delete();
+
+    val deleteResponseStatus = deleteResponse.getStatusCode();
+    assertEquals(deleteResponseStatus, OK);
+
+    val getResponse = initStringRequest().endpoint("/policies/%s/applications", policyId).get();
+
+    val getResponseStatus = getResponse.getStatusCode();
+    val getResponseJson = MAPPER.readTree(getResponse.getBody());
+
+    assertEquals(getResponseStatus, OK);
+    assertEquals(0, getResponseJson.get("count").asInt());
```

Comment:
i noticed this code is replicated 3 times. See if this can be refactored into a single method and then called in each flavour (user, application and group). Will make the tests much more readable.